src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs

```diff
@@ -311,86 +236,245 @@ private void InitializeUsingRsa(RSA rsa, string algorithm)
             return RSA.VerifyHash(HashAlgorithm.ComputeHash(bytes), signature, HashAlgorithmName, RSASignaturePadding);
         }
 #endif
+#endregion
 
+#region DESKTOP related code
 #if DESKTOP
+        internal byte[] DecryptWithRsaCryptoServiceProviderProxy(byte[] bytes)
+        {
+            return RsaCryptoServiceProviderProxy.Decrypt(bytes, _useRSAOeapPadding);
+        }
+
+        internal byte[] EncryptWithRsaCryptoServiceProviderProxy(byte[] bytes)
+        {
+            return RsaCryptoServiceProviderProxy.Encrypt(bytes, _useRSAOeapPadding);
+        }
+
+        private RSACryptoServiceProviderProxy RsaCryptoServiceProviderProxy { get; set; }
+
+        internal byte[] SignWithRsaCryptoServiceProviderProxy(byte[] bytes)
+        {
+            return RsaCryptoServiceProviderProxy.SignData(bytes, HashAlgorithm);
+        }
+
         private bool VerifyWithRsaCryptoServiceProviderProxy(byte[] bytes, byte[] signature)
         {
             return RsaCryptoServiceProviderProxy.VerifyData(bytes, HashAlgorithm, signature);
         }
 #endif
+#endregion
 
-        internal static bool VerifyFunctionNotFound(byte[] bytes, byte[] signature)
+#region NET45 'lightup' code
+        // the idea here is if a user has defined their application to target 4.6.1+ but some layer in the stack kicks down below, this code builds delegates
+        // for decrypting, encryption, signing and validating when we detect that the instance of RSA is RSACng and RSACng is supported by the framework.
+#if NET45
+        private byte[] DecryptNet45(byte[] bytes)
         {
-            // we should never get here, its a bug if we do.
-            throw LogHelper.LogExceptionMessage(new NotSupportedException(LogMessages.IDX10686));
+            if (_rsaDecrypt45Method == null)
+            {
+                // Decrypt(byte[] data, RSAEncryptionPadding padding)
+                Type[] encryptionTypes = { typeof(byte[]), _rsaEncryptionPaddingType };
+                MethodInfo encryptMethod = typeof(RSA).GetMethod(
+                    "Decrypt",
+                    BindingFlags.Public | BindingFlags.Instance,
+                    null,
+                    encryptionTypes,
+                    null);
+
+                Type delegateType = typeof(Func<,,,>).MakeGenericType(
+                            typeof(RSA),
+                            typeof(byte[]),
+                            _rsaEncryptionPaddingType,
+                            typeof(byte[]));
+
+                PropertyInfo prop;
+                if (_useRSAOeapPadding)
+                    prop = _rsaEncryptionPaddingType.GetProperty("OaepSHA1", BindingFlags.Static | BindingFlags.Public);
+                else
+                    prop = _rsaEncryptionPaddingType.GetProperty("Pkcs1", BindingFlags.Static | BindingFlags.Public);
+
+                Delegate openDelegate = Delegate.CreateDelegate(delegateType, encryptMethod);
+                _rsaDecrypt45Method = (rsaArg, bytesArg) =>
```

Comment:
This could be wrote to more than once (multiple threads calling Decrypt) Does that matter?

Reply:
the AsymmetricSignatureProvider maintains a pool of AsymmetricAdapters and they function as a singleton.